plugins/modules/lambda_layer_info.py

```diff
@@ -0,0 +1,201 @@
+- name: Retrieve layer versions
+  amazon.aws.lambda_layer_info:
+    name: blank-java-lib
+    compatible_architecture: x86_64
+
+# list latest versions of available layers
+- name: list latest versions for all layers
+  amazon.aws.lambda_layer_info:
+
+# list latest versions of available layers compatible with runtime python3.7
+- name: list latest versions for all layers
+  amazon.aws.lambda_layer_info:
+    compatible_runtime: python3.7
+'''
+
+RETURN = '''
+layers_versions:
+  description:
+  - The layers versions that exists.
+  returned: success
+  type: list
+  elements: dict
+  contains:
+    layer_arn:
+        description: The ARN of the layer.
+        returned: when C(name) is provided
+        type: str
+        sample: "arn:aws:lambda:eu-west-2:123456789012:layer:pylayer"
+    layer_version_arn:
+        description: The ARN of the layer version.
+        returned: if the layer version exists or has been created
+        type: str
+        sample: "arn:aws:lambda:eu-west-2:123456789012:layer:pylayer:2"
+    description:
+        description: The description of the version.
+        returned: I(state=present)
+        type: str
+    created_date:
+        description: The date that the layer version was created, in ISO-8601 format (YYYY-MM-DDThh:mm:ss.sTZD).
+        returned: if the layer version exists or has been created
+        type: str
+        sample: "2022-09-28T14:27:35.866+0000"
+    version:
+        description: The version number.
+        returned: if the layer version exists or has been created
+        type: int
+        sample: 1
+    compatible_runtimes:
+        description: A list of compatible runtimes.
+        returned: if it was defined for the layer version.
+        type: list
+        sample: ["python3.7"]
+    license_info:
+        description: The layer's software license.
+        returned: if it was defined for the layer version.
+        type: str
+        sample: "GPL-3.0-only"
+    compatible_architectures:
+        description: A list of compatible instruction set architectures.
+        returned: if it was defined for the layer version.
+        type: list
+'''
+
+try:
+    import botocore
+except ImportError:
+    pass  # Handled by AnsibleAWSModule
+
+from ansible.module_utils.common.dict_transformations import camel_dict_to_snake_dict
+
+from ansible_collections.amazon.aws.plugins.module_utils.core import AnsibleAWSModule
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import AWSRetry
+
+
+@AWSRetry.jittered_backoff()
+def _list_layer_versions(client, **params):
+    paginator = client.get_paginator('list_layer_versions')
+    return paginator.paginate(**params).build_full_result()
+
+
+@AWSRetry.jittered_backoff()
+def _list_layers(client, **params):
+    paginator = client.get_paginator('list_layers')
+    return paginator.paginate(**params).build_full_result()
+
+
+def list_layer_versions(module, lambda_client):
```

Comment:
I would suggest to write the function without the `module` parameter like this:

```python
def list_layer_versions(lambda_client, name, compatible_runtime=None, compatible_architecture=None):
    params = {"LayerName": name}
    if compatible_runtime:
        params["CompatibleRuntime"] = compatible_runtime
    if compatible_architecture:
        params["CompatibleArchitecture"] = compatible_architecture
    try:
        layer_versions = _list_layer_versions(lambda_client, **params)['LayerVersions']
        return [camel_dict_to_snake_dict(layer) for layer in layer_versions]
    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
        raise LambdaLayerInfoFailure(e, f"Unable to list layer versions for name {name}")
```

The exception would be similar to what we've got here: https://github.com/ansible-collections/amazon.aws/pull/1132/files#diff-a8d9a60fb28317dc33de7e91829936f988d3655dd43a558744e360e86601ca69R380-R384

When we pass the `module` object, it becomes harder to identify an unexpected side-effect of the function and this increase the testing complexity. Without this `module` parameter and a good control on the inputs, it will become easier to dedicate an unit-test to each of these functions.

The final code would look like this:

```python
class LambdaLayerInfoFailure(Exception):
    def __init__(self, original_e, user_message):
        self.original_e = original_e
        self.user_message = user_message
        super().__init__(self)


def list_layer_versions(lambda_client, name, compatible_runtime=None, compatible_architecture=None):
    params = {"LayerName": name}
    if compatible_runtime:
        params["CompatibleRuntime"] = compatible_runtime
    if compatible_architecture:
        params["CompatibleArchitecture"] = compatible_architecture
    try:
        layer_versions = _list_layer_versions(lambda_client, **params)['LayerVersions']
        return [camel_dict_to_snake_dict(layer) for layer in layer_versions]
    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
        raise LambdaLayerInfoFailure(e, f"Unable to list layer versions for name {name}")


def list_layers(lambda_client, name=None, compatible_runtime=None, compatible_architecture=None):

    params = {}
    if compatible_runtime:
        params["CompatibleRuntime"] = compatible_runtime
    if compatible_architecture:
        params["CompatibleArchitecture"] = compatible_architecture
    try:
        layers = _list_layers(lambda_client, **params)['Layers']
        layer_versions = []
        for item in layers:
            layer = {key: value for key, value in item.items() if key != "LatestMatchingVersion"}
            layer.update(item.get("LatestMatchingVersion"))
            layer_versions.append(camel_dict_to_snake_dict(layer))
        return layer_versions
    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
        raise LambdaLayerInfoFailure(e, f"Unable to list layers {params}")


def execute_module(module, lambda_client):

    operation = list_layers
    if module.params.get("name") is not None:
        operation = list_layer_versions

    try:
        module.exit_json(
            changed=False,
            layers_versions=operation(
                lambda_client,
                name=module.params.get("name"),
                compatible_runtime=module.params.get("compatible_runtime"),
                compatible_architecture=module.params.get("compatible_architecture")))
    except LambdaLayerInfoFailure as e:
        module.fail_json_aws(e.original_e, e.user_message)
```

Reply:
done, Thanks!